### PR TITLE
Fog of war

### PR DESCRIPTION
**Pull request by sybrid1982**
There's a lot of changes here, but essentially the gist of what happened here was:
1)  The View of the HexMap was pulled from the Data of the HexMap.  The View is now DisplayWorld, while the data stays in HexMap.  Relationship to Fog of War: made it easier to control what hexes are displayed to a player
2)  Display World is told at the start of a player's turn by the GameManager to display the map of hexes that the player has seen and otherwise displays black hexes for unseen tiles.  Relationship to Fog of War: Actually generated the fog of war
3)  Tests were added for the HexMap around the generation of neighbors.  Based on the results of those tests, the neighbor generation code was changed so that all the appropriate neighbors were being generated.  This was related to FogOfWar because currently a player reveals any adjacent hexes to his pawn, and some hexes that should have been flagged adjacent weren't.  All of the tests are green.
4)  When a player moves a pawn, that pawn reports its position and that is used to determine if any new hexes were revealed to the player.  Relationship to Fog of War:  Allows player to uncover new tiles.
5)  Camera controls were reworked.  This had nothing to do with FogOfWar, but they were irritating me so I did some more work on them.

### CODE REVIEW

**Comment by jzahnd** [comment]
Your indentations are off, some places it is 2 spaces and some it is 4. While that is a style issue, it is important to keep it consistent for readability sake. I'd fix that project wide.

File: Assets/Controllers/CameraMotionHandler.cs [hunk 1]
```diff
@@ -44,7 +48,7 @@ public void MoveToPosition(Vector3 position)
     {
         position.y = this.transform.position.y;
         position.z += -5;
-        this.transform.position = position;
+        newPosition = position;
```
Comment (jzahnd) [inline Assets/Controllers/CameraMotionHandler.cs:51]: You may want to pass this in as something like `currentPosition`, position and newPosition is a bit confusing to read. Also, the statically set -5 might be better off as a constant or config based setting somewhere.

File: Assets/Controllers/CameraMotionHandler.cs [hunk 2]
```diff
@@ -63,24 +67,43 @@ void CheckIfCameraMoved()
         }
     }
 
-    public void MoveCamera(int upDown, int leftRight)
+    public void SetMoveTargetPosition(int upDown, int leftRight)
     {
-        float newX = transform.position.x + (leftRight * Time.deltaTime * cameraMoveSpeed);
-        float newZ = transform.position.z + (upDown * Time.deltaTime * cameraMoveSpeed);
+        float newX = transform.position.x + (leftRight * Time.deltaTime);
```
Comment (jzahnd) [inline Assets/Controllers/CameraMotionHandler.cs:72]: Should this be a separate method that you pass the `transform.position` variable into and returns X and Z? It feels like this method is doing too much.

File: Assets/Controllers/CameraMotionHandler.cs [hunk 3]
```diff
@@ -63,24 +67,43 @@ void CheckIfCameraMoved()
...
         newY = Mathf.Clamp(newY, maxCameraZoomIn, maxCameraZoomOut);
 
-        transform.position = new Vector3(transform.position.x, newY, transform.position.z);
+        newPosition.y = newY;
+    }
+
+    void AdjustCameraAngle()
+    {
+        // Change camera angle
+        float lowZoom = maxCameraZoomIn + 3;
```
Comment (jzahnd) [inline Assets/Controllers/CameraMotionHandler.cs:94]: Abstract zoom mutations? This doesn't actually change the angle of the camera, that is done below.

File: Assets/Controllers/GameManager.cs [hunk 4]
```diff
@@ -6,47 +6,57 @@
...
 
 	// Use this for initialization
 	void Start () {
         selectedPawnComponent = null;
         selectedPawnDisplay = FindObjectOfType<UI_SelectedPawnDisplay>();
         uiPathDrawer = FindObjectOfType<UI_PathDrawer>();
+        startingNumberOfPlayers = 2;
 	}
 
     public void StartPressed()
```
Comment (jzahnd) [inline Assets/Controllers/GameManager.cs:28]: While this is an initialization method, it is starting to do a few too many things perhaps. You may want to think further about how you are handling this kick off.

File: Assets/Controllers/GameManager.cs [hunk 5]
```diff
@@ -106,7 +116,9 @@ void MovePawnVisual(Pawn pawn, Hex hex)
         //step two, move the pawn to the new location
         pawnTransform.position = hex.PositionFromCamera(Camera.main.transform.position, hexMap.NumRows(), hexMap.NumColumns());
         //Step three, set the pawn's parent to the new tile
-        pawnTransform.SetParent(hexMap.GetHexGOFromHex(hex).transform);
+        pawnTransform.SetParent(worldDisplay.GetHexGOFromHex(hex).transform);
+        //Step four, hey we moved a pawn, so let's go ahead and also change what tiles are visible
```
Comment (jzahnd) [inline Assets/Controllers/GameManager.cs:120]: Perhaps `worldDisplay.GetHexGOFromHex(hex).transform` should be it's own variable or method. This is hard to read inside of the `SetParent` call.

File: Assets/Controllers/GameManager.cs [hunk 6]
```diff
@@ -182,21 +194,27 @@ public Vector3 GetPawnPosition(Pawn pawn)
...
+        ClearSelectedPawn();
+        SetActivePlayer(players[playerIndex]);
+        activePlayer.StartTurn();
+        GetFirstPawnForActivePlayerPositionAndFocusCamera();
+    }
 
-        int playerIndex = players.FindIndex( x => x == activePlayer);
+    private int GetNextPlayerIndex()
+    {
+        int playerIndex = players.FindIndex(x => x == activePlayer);
```
Comment (jzahnd) [inline Assets/Controllers/GameManager.cs:206]: Is there value in finding the index of any player? Another abstraction opportunity where you pass in a player (in this case `activePlayer`) and it returns an index.

File: Assets/Controllers/MouseInputManager.cs [hunk 7]
```diff
@@ -7,18 +7,23 @@ public class MouseInputManager : MonoBehaviour {
 
     [SerializeField]
     int hexLayer;
+    [SerializeField]
+    float maxTimeAllowedToDoubleClick = 0.5f;
```
Comment (jzahnd) [inline Assets/Controllers/MouseInputManager.cs:11]: No.

File: Assets/Controllers/MouseInputManager.cs [hunk 8]
```diff
@@ -74,10 +84,19 @@ private void EvaluateLeftMouseClick()
                 else if (hit.collider.GetComponentInParent<HexComponent>() != null)
                 {
                     HexComponent hexComponent = hit.collider.GetComponentInParent<HexComponent>();
-                    gameManager.HexClicked(hexComponent.hex);
+                    if(lastHexClicked == hexComponent.hex && timeSinceLastClick <= maxTimeAllowedToDoubleClick)
```
Comment (jzahnd) [inline Assets/Controllers/MouseInputManager.cs:87]: Potential conditional abstraction.

File: Assets/Editor/HexMapTests.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    public void AssertHexAtQ5R5HasLowerRightNeighbor()
+    {
+        HexMap hexMap = ScriptableObject.CreateInstance<HexMap>();
+        hexMap.GenerateMap();
+        Hex hex = hexMap.GetHexAt(5, 5);
+        Hex neighbor = hex.GetNeighbor(Direction.LOWER_RIGHT);
+        Assert.AreNotEqual(null, neighbor);
+    }
+
+    //The next six tests take a hex on the left edge of the map and check that it has neighbors
```
Comment (jzahnd) [inline Assets/Editor/HexMapTests.cs:102]: Spaces or multiline comments?

File: Assets/HexMap.cs [hunk 10]
```diff
@@ -129,16 +101,17 @@ void GenerateNeighbors(Hex h)
         //When a new hex is created there are three easy places to check for neighbors
         //To the left of the hex, the bottom-left of the hex, and bottom-right
         //neighbor to the left
-        Vector2[] potentialNeighborCoords = {   new Vector2 (h.Q, h.R - 1),          //left
-                                                new Vector2 (h.Q - 1, h.R),         //lower left
-                                                new Vector2 (h.Q - 1, h.R + 1),};    //lower right
+        Vector2[] potentialNeighborCoords = {   new Vector2 (h.Q - 1, h.R + 1),     //upper left
```
Comment (jzahnd) [inline Assets/HexMap.cs:104]: An abstracted method that returns a vector?

File: Assets/HexMap.cs [hunk 11]
```diff
@@ -215,7 +188,9 @@ public Hex GetAcceptableStartPosition()
             int impassibleTiles = 0;
             foreach (Hex h in neighbors)
             {
-                if (h.GetTerrain() == Terrain.MOUNTAIN || h.GetTerrain() == Terrain.WATER)
+                if (h == null)
+                    Debug.LogWarning("Null neighbor, something is borked");
```
Comment (jzahnd) [inline Assets/HexMap.cs:192]: No.

File: Assets/HexMap.cs [hunk 12]
```diff
@@ -215,7 +188,9 @@ public Hex GetAcceptableStartPosition()
...
+                else if (h.GetTerrain() == Terrain.MOUNTAIN || h.GetTerrain() == Terrain.WATER)
```
Comment (jzahnd) [inline Assets/HexMap.cs:193]: Abstracted method that you pass in the terrain and it tells returns a type, then use case statement.

File: Assets/HexMap.cs [hunk 13]
```diff
@@ -226,96 +201,54 @@ public Hex GetAcceptableStartPosition()
...
     
-    public void UpdateHexVisuals()
+    public Hex[] GetHexesWithRadiusOf(Hex centerHex, int radius)
     {
-        for (int column = 0; column < NumColumns(); column++)
+        List<Hex> results = new List<Hex>();
+        for (int dx = -radius; dx < radius; dx++)
         {
-            for (int row = 0; row < NumRows(); row++)
+            for (int dy = Mathf.Max(-radius, -dx - radius); dy < Mathf.Min(radius, -dx + radius); dy++)
```
Comment (jzahnd) [inline Assets/HexMap.cs:210]: Nested for loop, complicated calculation with no comment (again I'd look at a dedicated calculation solution).

File: Assets/HexMap.cs [hunk 14]
```diff
@@ -226,96 +201,54 @@ public Hex GetAcceptableStartPosition()
...
+            walkableHexes.Add(hex);
+            if (hex.Moisture >= MoisturePlains)
+                hex.SetTerrain(Terrain.GRASS);
+            else
+                hex.SetTerrain(Terrain.DESERT);
+        }
+        else
+            hex.SetTerrain(Terrain.WATER);
+    }
+
```
Comment (jzahnd) [inline Assets/HexMap.cs:241]: Is this a WIP? The logic needs to go in its own manager / object or the like. Keep track of this kind of flow manually is painful.

File: Assets/HexMap.cs [hunk 15]
```diff
@@ -226,96 +201,54 @@ public Hex GetAcceptableStartPosition()
...
+    protected void CreateTerrainForHexes()
+    {
+        //Done setting the numbers so generate the terrain
+        for (int column = 0; column < NumColumns(); column++)
+        {
+            for (int row = 0; column < NumRows(); column++)
```
Comment (jzahnd) [inline Assets/HexMap.cs:247]: Nested for loop. Perhaps make a method that interacts on a column

File: Assets/Pathfinding/HexGraph.cs [hunk 16]
```diff
@@ -9,10 +9,9 @@ public HexGraph(HexMap hexMap)
     {
         nodes = new Dictionary<Hex, Path_Node<Hex>>();
 
-        //loop through all the hexes of the world
-        //create a node for each hex
-        //no nodes for unwalkable hexes
-        //TODO: Implement existence of non-walkable hexes
+        /*loop through all the hexes of the world
```
Comment (jzahnd) [inline Assets/Pathfinding/HexGraph.cs:12]: Add space to me!

File: Assets/UI/UI_PathDrawer.cs [hunk 17]
```diff
@@ -80,7 +80,7 @@ private void CreateMarkerAtHex(Hex hexToMark, Sprite spriteToMark)
         Vector3 markerPosition = hexToMark.PositionFromCamera(Camera.main.transform.position);
         markerPosition.y += downwardAdjustOfMarkers;
         //need to adjust upward for hills
-        if (hexToMark.GetTerrain() == Terrain.HILLS)
+        if (hexToMark.GetTerrain() == Terrain.ARID_HILLS || hexToMark.GetTerrain() == Terrain.GRASSY_HILLS)
```
Comment (jzahnd) [inline Assets/UI/UI_PathDrawer.cs:83]: Same as other comment about returning a type.

File: Assets/WorldDisplay.cs [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    public Mesh MeshFlat;
+    public Mesh MeshHill;
+    public Mesh MeshMountain;
+
+    public void Initialize(HexMap hexMap)
+    {
+        hexToHexGOMap = new Dictionary<Hex, GameObject>();
+        for(int q = 0; q < hexMap.NumColumns(); q++)
+        {
+            for (int r = 0; r < hexMap.NumRows(); r++)
```
Comment (jzahnd) [inline Assets/WorldDisplay.cs:28]: Same as other nested for loop comment.

File: Assets/WorldDisplay.cs [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        }
+        foreach (Hex revealedHex in player.GetRevealedButNotVisibleHexes())
+        {
+            UpdateHexVisuals(revealedHex);
+        }
+    }
+
+    private void GenerateHexGO(Hex h)
+    {
+        GameObject hexGO = (GameObject)Instantiate(hexPrefab,
```
Comment (jzahnd) [inline Assets/WorldDisplay.cs:50]: Can this be moved into a factory of some kind?

File: Assets/Controllers/GameManager.cs [hunk 20]
```diff
@@ -208,7 +226,24 @@ public void GetFirstPawnForActivePlayerPositionAndFocusCamera()
...
+    }
+
+    public WorldDisplay GetWorldDisplay()
+    {
+        return worldDisplay;
+    }
+
+    void SuggestEndingTurn()
+    {
+        print("You have no pawns left with turns!");
```
Comment (jzahnd) [inline Assets/Controllers/GameManager.cs:247]: Should this be a user notification? Or maybe it is? Does this print to a console?

File: Assets/Pawns/Pawn.cs [hunk 21]
```diff
@@ -43,23 +44,26 @@ public Pawn(string pawnType, int passedMovement)
...
-    public delegate void PawnMovesToNewHex(Pawn pawn, Hex hex);
+    public delegate void PawnMovesToNewHex(Pawn pawn);
     public event PawnMovesToNewHex pawnMoveEvent;
 
     public void SetMyHex(Hex newHex)
     {
+        if (myHex != null)
+            myHex.RemovePawn(this);
+        
+        Hex oldHex = myHex;
```
Comment (jzahnd) [inline Assets/Pawns/Pawn.cs:63]: oldHex? can this naming be updated to be more expressive?

File: Assets/Player.cs [hunk 22]
```diff
@@ -29,14 +34,31 @@ protected set
...
         myPawns = new List<Pawn>();
+        revealedHexes = new HashSet<Hex>();
+        currentlyVisibleHexes = new HashSet<Hex>();
         playerName = name;
     }
 
     public void StartTurn()
     {
         if (myPawns.Count > 0)
             foreach (Pawn p in myPawns)
```
Comment (jzahnd) [inline Assets/Player.cs:45]: Abstract into a method and send into a collection?

File: Assets/Player.cs [hunk 23]
```diff
@@ -74,4 +113,37 @@ public Pawn GetPawnWithRemainingMove()
...
+    {
+        //These hexes will ideally have some sort of 'fog of war' element
+        //applied to them.  For now maybe just make their material half alpha?
+        List<Hex> returnedHexes = new List<Hex>();
+
+        foreach(Hex h in revealedHexes)
+        {
+            if (currentlyVisibleHexes.Contains(h))
+            {
+                //Then don't add it to the list we're building
```
Comment (jzahnd) [inline Assets/Player.cs:132]: Do something here.

File: Assets/WorldDisplay.cs [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+            case Terrain.DESERT:
+                mr.material = MatDesert;
+                mf.mesh = MeshFlat;
+                break;
+            case Terrain.WATER:
+                mr.material = MatOcean;
+                mf.mesh = MeshWater;
+                break;
+            default:
+                Debug.LogError("Unknown terrain " + h.GetTerrain().ToString() + " sent to WorldDisplay");
```
Comment (jzahnd) [inline Assets/WorldDisplay.cs:114]: No.

File: Assets/Controllers/MouseInputManager.cs (hunk 7 above)
Comment (sybrid1982) [reply Assets/Controllers/MouseInputManager.cs:11]: So, a serialized field lets me see and adjust it in the editor.  I haven't decided how a double click 'should feel' yet, so I have it set here like this.  Once I've picked a value I like, I'll either make this a static, or I'll make that a default somewhere and have an option to change the time delay that allows a double click.

File: Assets/HexMap.cs (hunk 11 above)
Comment (sybrid1982) [reply Assets/HexMap.cs:192]: Could you explain what you don't like about this?  Is it the debug warning, the checking for null, or what?  Would you prefer I wrapped this in a try/catch type statement?

Comment (jzahnd) [reply Assets/HexMap.cs:192]: The debug warning. There is sometimes a smell in checking for null, but that seems too nitpicky at this stage. I would think you should be throwing and handling an error further up the chain and logging it, but maybe not. I am not sure of the best practice in a unity project.

File: Assets/Controllers/MouseInputManager.cs (hunk 7 above)
Comment (jzahnd) [reply Assets/Controllers/MouseInputManager.cs:11]: This is another one where I don't have enough understanding of the framework mechanisms to counter (or agree). I'd think a config file would let you do the same thing but maybe that is IDE specific.

**Pull request merged**